### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by andrew-kuzmin**

### CODE REVIEW

File: src/main/java/com/phoenixoft/teambalanceapp/user/service/UserService.java [hunk 1]
```diff
@@ -39,6 +60,72 @@ public User update(Long userId, UserRequestDto dto) {
...
+        String token = UUID.randomUUID().toString();
+
+        PasswordResetToken passwordResetToken = new PasswordResetToken();
+        passwordResetToken.setToken(token);
+        passwordResetToken.setUser(user);
+        passwordResetToken.setExpiryDate(LocalDateTime.now().plusHours(2));
+        passwordTokenRepository.save(passwordResetToken);
+
+        String mailContent = "To complete the password reset process, please click here: "
+                + "http://localhost:5000/confirm-reset?token=" + token;
```
Comment (JackKarichkovskiy) [inline src/main/java/com/phoenixoft/teambalanceapp/user/service/UserService.java]: Please extract host to external config (application.properties)

File: src/main/java/com/phoenixoft/teambalanceapp/util/AuthenticationUtils.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthenticationUtils {
+
+    public Long getLoggedInUserId() {
+        SecurityContext securityContext = SecurityContextHolder.getContext();
+        CustomUser user;
```
Comment (JackKarichkovskiy) [inline src/main/java/com/phoenixoft/teambalanceapp/util/AuthenticationUtils.java]: Why should be declared here?

File: src/main/java/com/phoenixoft/teambalanceapp/util/AuthenticationUtils.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        if (null != securityContext.getAuthentication()){
```
Comment (JackKarichkovskiy) [inline src/main/java/com/phoenixoft/teambalanceapp/util/AuthenticationUtils.java]: Perfect match to use Optional

File: src/main/java/com/phoenixoft/teambalanceapp/util/SendGridMailService.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+        Request request = new Request();
+        try {
+            request.setMethod(Method.POST);
+            request.setEndpoint("mail/send");
+            request.setBody(mail.build());
+            Response response = this.sendGrid.api(request);
+            sendGrid.api(request);
+        } catch (IOException ex) {
+
```
Comment (JackKarichkovskiy) [inline src/main/java/com/phoenixoft/teambalanceapp/util/SendGridMailService.java]: It's a bad smell to empty catch block. Please log exception.

File: build.gradle [hunk 5]
```diff
@@ -24,10 +24,10 @@ repositories {
 
 dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-actuator'
-    implementation 'org.springframework.boot:spring-boot-starter-data-rest'
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
     implementation 'org.springframework.boot:spring-boot-starter-security'
+    compile group: 'com.sendgrid', name: 'sendgrid-java', version: '4.4.1'
```
[the remark this answers is not in the document]
Comment (andrew-kuzmin) [reply build.gradle]: done

File: src/main/java/com/phoenixoft/teambalanceapp/controller/GroupController.java [hunk 6]
```diff
@@ -67,7 +75,17 @@ public GroupResponseDto addMember(@PathVariable Long groupId, @PathVariable Long
...
+        UserDetails userDetails = (UserDetails) authentication.getPrincipal();
+        groupService.checkAdminPermissions(userDetails, groupId);
         groupService.deleteMember(groupId, userId);
     }
+
+    @PostMapping(path = "/{groupId}/members/{userId}/addAdminPrivileges")
+    public void assignUserToGroupAdmins(@PathVariable Long groupId, @PathVariable Long userId, Authentication authentication) {
+        UserDetails userDetails = (UserDetails) authentication.getPrincipal();
+        groupService.checkAdminPermissions(userDetails, groupId);
+        // todo refactor to check group existence first
```
[the remark this answers is not in the document]
Comment (andrew-kuzmin) [reply src/main/java/com/phoenixoft/teambalanceapp/controller/GroupController.java]: done

File: src/main/java/com/phoenixoft/teambalanceapp/group/service/GroupService.java [hunk 7]
```diff
@@ -74,4 +90,50 @@ public void deleteMember(Long groupId, Long userId) {
...
+            Role userRole = new Role();
+            userRole.setName(USER_ROLE_PREFIX + groupId);
+            roleRepository.save(userRole);
+
+            creator.getRoles().addAll(new HashSet<>(Arrays.asList(adminRole, userRole)));
+            userRepository.save(creator);
+        } else {
+            throw new ServiceException("Could not create new group.", HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (andrew-kuzmin) [reply src/main/java/com/phoenixoft/teambalanceapp/group/service/GroupService.java]: I will move it to a separate service later.

File: src/main/java/com/phoenixoft/teambalanceapp/group/service/GroupService.java [hunk 8]
```diff
@@ -74,4 +90,50 @@ public void deleteMember(Long groupId, Long userId) {
...
+
+    public void assignUserRoleToUser(Long groupId, Long userId) {
+        String processingGroupAdminRole = USER_ROLE_PREFIX + groupId;
+        Role existingAdminRole = roleRepository.findByName(processingGroupAdminRole)
+                .orElseThrow(() -> new ResourceNotFoundException("User role for the group not found"));
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new ResourceNotFoundException("User not found: " + userId));
+        user.getRoles().add(existingAdminRole);
+        userRepository.save(user);
+    }
```
[the remark this answers is not in the document]
Comment (andrew-kuzmin) [reply src/main/java/com/phoenixoft/teambalanceapp/group/service/GroupService.java]: I think we should define what to do in kind of situation. You are right from the side of correctness and clarity. But if we have a goal to launch asap then we could skip similar issues.

[the remark this answers is not in the document]
Comment (andrew-kuzmin) [reply src/main/java/com/phoenixoft/teambalanceapp/group/service/GroupService.java]: @JackKarichkovskiy

File: src/main/java/com/phoenixoft/teambalanceapp/group/service/GroupService.java [hunk 9]
```diff
@@ -74,4 +90,50 @@ public void deleteMember(Long groupId, Long userId) {
...
+        String processingGroupAdminRole = ADMIN_ROLE_PREFIX + groupId;
+        Set<SimpleGrantedAuthority> authorities = (Set<SimpleGrantedAuthority>) userDetails.getAuthorities();
+        boolean matchAdminRole = authorities.stream().anyMatch(simpleGrantedAuthority -> simpleGrantedAuthority.getAuthority().equals(processingGroupAdminRole));
+        if (!matchAdminRole) {
+            throw new AccessDeniedException("Admin privileges required.");
+        }
+    }
+
+    private void createAndAssignGroupRoles(Long groupId, User creator) {
+        if (groupId != null) {
```
[the remark this answers is not in the document]
Comment (andrew-kuzmin) [reply src/main/java/com/phoenixoft/teambalanceapp/group/service/GroupService.java]: I didn't get you.

File: src/main/java/com/phoenixoft/teambalanceapp/user/repository/UserRepository.java [hunk 10]
```diff
@@ -1,9 +1,15 @@
...
 
+import java.util.Optional;
+
 @Repository
-public interface UserRepository extends CrudRepository<User, Long> {
+public interface UserRepository extends JpaRepository<User, Long> {
+
+    Optional<User> findByEmail(String email);
+
+    Boolean existsByEmail(String email);
```
[the remark this answers is not in the document]
Comment (andrew-kuzmin) [reply src/main/java/com/phoenixoft/teambalanceapp/user/repository/UserRepository.java]: done

File: src/main/java/com/phoenixoft/teambalanceapp/user/service/UserService.java (hunk 1 above)
Comment (andrew-kuzmin) [reply src/main/java/com/phoenixoft/teambalanceapp/user/service/UserService.java]: The feature is not implemented yet.

File: src/main/java/com/phoenixoft/teambalanceapp/util/AuthenticationUtils.java (hunk 2 above)
Comment (andrew-kuzmin) [reply src/main/java/com/phoenixoft/teambalanceapp/util/AuthenticationUtils.java]: It is a concept.

File: src/main/java/com/phoenixoft/teambalanceapp/util/SendGridMailService.java (hunk 4 above)
Comment (andrew-kuzmin) [reply src/main/java/com/phoenixoft/teambalanceapp/util/SendGridMailService.java]: The feature is not implemented yet.

**Pull request merged**